### PR TITLE
Using app config (if it exists) to resolve scheduled activities.

### PR DESCRIPTION
**Pull request by alxdarksage**
This finishes app configs by getting the app config for the request access scheduled activities, and referring to it before looking up the most recent schema/survey. Logic to resolve schedules was moved into separate class to make it easier to test (ReferenceResolver). I did not remove tests of reference resolution in the other tests, let me know if we should do this.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/services/AppConfigService.java [hunk 1]
```diff
@@ -78,10 +78,12 @@ public AppConfig getAppConfigForUser(CriteriaContext context) {
...
         // Should have matched one and only one app config.
-        // The goal of the following code is not to introduce production exceptions when changing app configs.
         if (matches.isEmpty()) {
-            // If there are no matches, return the "default" app config (TBD: for now, throw exception)
-            throw new EntityNotFoundException(AppConfig.class);
+            if (throwException) {
+                throw new EntityNotFoundException(AppConfig.class);    
+            } else {
+                return null;
+            }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AppConfigService.java:86]: Is it worth adding a unit test for throwException=false just for branch coverage?

File: test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        resolver.resolve(scheduledActivity);
+        
+        verify(compoundActivityDefinitionService, times(1)).getCompoundActivityDefinition(STUDY_ID, TASK_ID);
+    }
+    
+    @Test
+    public void compoundActivityFullyResolvedFromAppConfig() {
+        schemaReferences.put(SCHEMA_ID, RESOLVED_SCHEMA_REF);
+        surveyReferences.put(SURVEY_GUID, RESOLVED_SURVEY_REF);
+        scheduledActivity.setActivity(activityBuilder.withCompoundActivity(UNRESOLVED_COMPOUND_ACTIVITY).build());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java:252]: I think it's worth using the skinny ref here (and probably in the next 3 tests). This will test that we recursively resolve schemas and surveys inside a compound activity ref.

Alternatively, one extra test cases that specifically tests this case.

File: test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        
+        resolver.resolve(scheduledActivity);
+        
+        assertNull(scheduledActivity.getActivity().getTask().getSchema().getRevision());
+    }
+    
+    @Test
+    public void unresolvableCompoundActivityReturnedAsIs() {
+        scheduledActivity.setActivity(activityBuilder.withCompoundActivity(COMPOUND_ACTIVITY_SKINNY_REF).build());
+        when(compoundActivityDefinitionService.getCompoundActivityDefinition(STUDY_ID, TASK_ID)).thenThrow(new EntityNotFoundException(CompoundActivityDefinition.class));
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java:358]: Should we also test a compound activity with unresolvable schemas and surveys?

File: app/org/sagebionetworks/bridge/services/ReferenceResolver.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            SurveyReference resolvedSurveyRef = resolveSurvey(surveyRef);
+
+            if (resolvedSurveyRef != null && !resolvedSurveyRef.equals(surveyRef)) {
+                resolvedActivity = new Activity.Builder().withActivity(activity).withSurvey(resolvedSurveyRef).build();
+            }
+        } else if (activityType == ActivityType.TASK) {
+            TaskReference taskRef = activity.getTask();
+            SchemaReference schemaRef = taskRef.getSchema();
+
+            if (schemaRef != null) {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ReferenceResolver.java:89]: Is it worth adding a test case for a taskRef without a schemaRef, for branch coverage?

File: app/org/sagebionetworks/bridge/services/ScheduledActivityService.java [hunk 5]
```diff
@@ -366,189 +358,32 @@ private int byteLength(JsonNode node) {
...
 
         List<SchedulePlan> plans = schedulePlanService.getSchedulePlans(context.getCriteriaContext().getClientInfo(),
                 context.getCriteriaContext().getStudyIdentifier());
-
+        
+        AppConfig appConfig = appConfigService.getAppConfigForUser(context.getCriteriaContext(), false);
+        Map<String, SurveyReference> surveyReferences = (appConfig == null) ? ImmutableMap.of()
+                : Maps.uniqueIndex(appConfig.getSurveyReferences(), SurveyReference::getGuid);
+        Map<String, SchemaReference> schemaReferences = (appConfig == null) ? ImmutableMap.of()
+                : Maps.uniqueIndex(appConfig.getSchemaReferences(), SchemaReference::getId);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ScheduledActivityService.java:370]: Is it worth adding a minimal test to ScheduledActivityServiceResolveLinksTest with non-null AppConfig and non-empty reference maps, just to make sure we're constructing the maps correctly?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Looks good. All my notes are test notes.

I'm okay with leaving the tests in ScheduledActivityServiceResolveLinksTest as they are, since they run pretty fast as is.

I'd also be okay with removing most of the tests in ScheduledActivityServiceResolveLinksTest, since they are redundant. As long as we leave in minimal tests to test both the AppConfig null and non-null cases.

File: app/org/sagebionetworks/bridge/services/AppConfigService.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/AppConfigService.java:86]: Added test for null case.

File: test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java (hunk 2 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java:252]: I've added a test of a reference compound activity that retrieves a definition where the references aren't resolved, and verify that those in turn are verified.

File: test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java (hunk 3 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/ReferenceResolverTest.java:358]: Added test where compound activity def is returned but the references are broken (not found by the services).

File: app/org/sagebionetworks/bridge/services/ReferenceResolver.java (hunk 4 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ReferenceResolver.java:89]: added

File: app/org/sagebionetworks/bridge/services/ScheduledActivityService.java (hunk 5 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ScheduledActivityService.java:370]: Added a test where there's an AppConfig, verify the app configs references are used to resolve the scheduled activities, so the map should be correct.

**Review by DwayneJengSage — APPROVED** [review]
